assets/src/blocks/Counter/CounterFrontend.js

```diff
@@ -0,0 +1,123 @@
+import { Component, Fragment } from '@wordpress/element';
+
+export class CounterFrontend extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      remaining: 0,
+      completed: 0
+    };
+
+    this.calculateRemaining = this.calculateRemaining.bind(this);
+    this.getCounterText = this.getCounterText.bind(this);
+  }
+
+  componentDidMount() {
+    // Calculate completed and remaining values depending on props
+    this.calculateRemaining();
+  }
+
+  componentDidUpdate({ target: prevTarget, completed: prevCompleted, completed_api: prevCompletedApi }) {
+    // Update completed and remaining values depending on props
+    const { target, completed, completed_api } = this.props;
+    if (target !== prevTarget || completed !== prevCompleted || completed_api !== prevCompletedApi) {
+      this.calculateRemaining();
+    }
+  }
+
+  calculateRemaining() {
+    const { completed_api } = this.props;
+    const target = this.props.target || 0;
+    let completed = this.props.completed || 0;
+    let remaining = 0;
+    if (completed_api) {
+      fetch(completed_api)
+        .then(response => response.json())
+        .then(({ unique_count }) => {
+          if (unique_count) {
+            completed = unique_count;
+            this.setState({
+              completed,
+              remaining: target - completed
+            });
+          }
+        });
+    } else if (target > 0 || completed > 0) {
+      remaining = target - completed;
+      this.setState({ remaining, completed });
+    }
+  }
+
+  getCounterText() {
+    const { text, target } = this.props;
+    const { remaining, completed } = this.state;
+
+    const COUNTER_TEXT = {
+      '%completed%': `<span class="counter-target">${completed}</span>`,
+      '%target%': `<span class="counter-target">${target || 0}</span>`,
+      '%remaining%': `<span class="counter-target">${remaining}</span>`
+    };
+
+    return text.replace(/%completed%|%target%|%remaining%/gi, match => COUNTER_TEXT[match]);
+  }
+
+  render() {
+    const {
+      className,
+      title,
+      description,
+      text,
+      target,
+      isEditing
+    } = this.props;
+
+    const { completed } = this.state;
+
+    let style = 'plain';
+    if (className) style = className.split('is-style-')[1];
+    let arcLength = 31.5;
+
+    const percent = target > 0 ? Math.round(completed / target * 100) : 0;
```

Comment:
Same here, percent shouldn't go over 100 (the arc starts emptying with completed > target), so `Math>max(target > 0 ? Math.round(completed / target * 100) : 0, 100)` maybe ?

Reply:
It should be `Math.min` though, right? 🤔

Reply:
Right, don't believe anything I say when maths are involved 🙈